pkg/supportbundle/supportbundle.go

```diff
@@ -0,0 +1,236 @@
+	}()
+
+	// Gather runtime metadata.
+	ut, err := uptime.Get()
+	if err != nil {
+		return nil, err
+	}
+	m := Metadata{
+		BuildVersion: build.Version,
+		OS:           runtime.GOOS,
+		Architecture: runtime.GOARCH,
+		Uptime:       ut.Seconds(),
+		Payload:      map[string]interface{}{"enabled-features": enabledFeatures},
+	}
+	meta, err := yaml.Marshal(m)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal support bundle metadata: %s", err)
+	}
+	// Gather current configuration.
+	config, err := yaml.Marshal(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal config: %s", err)
+	}
+
+	var httpClient http.Client
+	httpClient.Transport = &http.Transport{DialContext: dialContext}
+	// Gather Agent's own metrics.
+	resp, err := httpClient.Get("http://" + srvAddress + "/metrics")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get internal Agent metrics: %s", err)
+	}
+	agentMetrics, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read internal Agent metrics: %s", err)
+	}
+
+	// Collect the Agent metrics instances and target statuses.
+	resp, err = httpClient.Get("http://" + srvAddress + "/agent/api/v1/metrics/instances")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get internal Agent metrics: %s", err)
+	}
+	agentMetricsInstances, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read internal Agent metrics: %s", err)
+	}
+	resp, err = httpClient.Get("http://" + srvAddress + "/agent/api/v1/metrics/targets")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Agent metrics targets: %s", err)
+	}
+	agentMetricsTargets, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read Agent metrics targets: %s", err)
+	}
+
+	// Collect the Agent's logs instances and target statuses.
+	resp, err = httpClient.Get("http://" + srvAddress + "/agent/api/v1/logs/instances")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Agent logs instances: %s", err)
+	}
+	agentLogsInstances, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read Agent logs instances: %s", err)
+	}
+	// TODO(@tpaschalis) Add back after grafana/agent@2175 is resolved, as it
+	// currently results in a panic.
+	// resp, err = http.DefaultClient.Get("http://" + srvAddress + "/agent/api/v1/logs/targets")
+	// if err != nil {
+	// 	return fmt.Errorf("failed to get  Agent logs targets: %s", err)
+	// }
+	// agentLogsTargets, err := io.ReadAll(resp.Body)
+	// if err != nil {
+	// 	return fmt.Errorf("failed to read internal Agent metrics: %s", err)
+	// }
+
+	// Export pprof data.
+	var (
+		cpuBuf       bytes.Buffer
+		heapBuf      bytes.Buffer
+		goroutineBuf bytes.Buffer
+		blockBuf     bytes.Buffer
+		mutexBuf     bytes.Buffer
+	)
+	err = pprof.StartCPUProfile(&cpuBuf)
+	if err != nil {
+		return nil, err
+	}
+	deadline, _ := ctx.Deadline()
+	time.Sleep(time.Until(deadline) - 100*time.Millisecond)
```

Comment:
This looks like a magic value. Do we want to document why 100ms?

Reply:
Sorry for not responding to this before.

The idea was to spend _most_ of the given `duration` to let the profiles gather some data, and then give it a small amount of time to write those profiles and the result of the bundle.

Do you have any better way in mind to do this? Otherwise, I'll just add a comment as you suggested.